parser.hh

```diff
@@ -388,6 +404,9 @@ struct ParserPosition
 	///iterator into the input.
 	Input::iterator it;
 
+	///user-meaningful filename.
+	std::string filename;
```

Comment:
Ok, I've tried another tack. If you object to the addition of a `NoInput` class, we can switch `Input::iterator::input()` to return a pointer instead (which might be nullptr).